### ISSUE TITLE
Model size on disk bigger than what was reported by lq.models.summary()

### ISSUE THREAD

**Issue by rameshKrSah** [larq/compute-engine#669]
I am running the into_to_cnns [notebook](https://docs.larq.dev/larq/tutorials/mnist/) and the size of the binary model reported by `lq.models.summary()` is 13.48 KiB but the saved model on disk is 166.1 KB. 

![image](https://user-images.githubusercontent.com/21354640/122864088-d26c1680-d2d8-11eb-8737-787e4fe627f9.png)

![image](https://user-images.githubusercontent.com/21354640/122864108-de57d880-d2d8-11eb-8e5e-36887e678fcc.png)

After changing the `QuantDense` layers to `QuantConv2d` the model size decreased substantially but still not what was reported by `lq.models.summary()`

![image](https://user-images.githubusercontent.com/21354640/122864207-0e9f7700-d2d9-11eb-8c0f-2d5c9a8f2c9b.png)

I was wondering whether it is possible to obtain a model with size same as that reported by `lq.models.summary()`.

Thank you!

I ran the into_to_cnns notebook on Google colab. 

```
# All quantized layers except the first will use the same options
kwargs = dict(input_quantizer="ste_sign",
              kernel_quantizer="ste_sign",
              kernel_constraint="weight_clip")

model = tf.keras.models.Sequential()

# In the first layer we only quantize the weights and not the input
model.add(lq.layers.QuantConv2D(32, (3, 3),
                                kernel_quantizer="ste_sign",
                                kernel_constraint="weight_clip",
                                use_bias=False,
                                input_shape=(28, 28, 1)))
model.add(tf.keras.layers.MaxPooling2D((2, 2)))
model.add(tf.keras.layers.BatchNormalization(scale=False))

model.add(lq.layers.QuantConv2D(64, (3, 3), use_bias=False, **kwargs))
model.add(tf.keras.layers.MaxPooling2D((2, 2)))
model.add(tf.keras.layers.BatchNormalization(scale=False))

model.add(lq.layers.QuantConv2D(64, (3, 3), use_bias=False, **kwargs))
model.add(tf.keras.layers.BatchNormalization(scale=False))
model.add(tf.keras.layers.Flatten())

model.add(tf.keras.layers.Reshape((1, 1, -1)))
model.add(lq.layers.QuantConv2D(64, kernel_size=(1, 1), **kwargs))
model.add(tf.keras.layers.Reshape((-1,)))

# model.add(lq.layers.QuantDense(64, use_bias=False, **kwargs))
model.add(tf.keras.layers.BatchNormalization(scale=False))


# model.add(lq.layers.QuantDense(10, use_bias=False, **kwargs))

model.add(tf.keras.layers.Reshape((1, 1, -1)))
model.add(lq.layers.QuantConv2D(10, kernel_size=(1, 1), **kwargs))
model.add(tf.keras.layers.Reshape((-1,)))

model.add(tf.keras.layers.BatchNormalization(scale=False))
model.add(tf.keras.layers.Activation("softmax"))

```

**Comment by lgeiger** [comment]
> After changing the QuantDense layers to QuantConv2d the model size decreased substantially but still not what was reported by lq.models.summary()

Unfortunately this is currently expected, since fully connected layers are currently not accelerated by larq compute engine. For more information please checkout #53

> I was wondering whether it is possible to obtain a model with size same as that reported by lq.models.summary().

There are a few subtleties here. `lq.models.summary()` gives you an estimate of the theoretical expected models size that is independent of the inference engine used.
Unfortunately this estimate might be off in some cases due to some reasons:
- The summary might not be aware of all the fusion rules that the inference engine supports or might not recognise all model features correctly which means it could end up with a wrong estimate. We would be super happy for PRs improving these estimates in the model summary [here](https://github.com/larq/larq/blob/master/larq/models.py).
- In larq compute engine, binary weights are packed across the channel dimension into `int32` values so there might be some padding required that will increase the final model size compared to the theoretical estimate in cases where the number of filters are not multiples of 32. We have [docs for this available](https://docs.larq.dev/compute-engine/model_optimization_guide/) explaining these compute engine specifics in more detail.
- There is a small overhead in the TFLite flatbuffer that stores the computational graph and some metadata like quantization stats that are not accounted for in the model summary although this overhead is probably only noticeable in very small models.

Does this answer your question?

**Comment by rameshKrSah** [comment]
Thank you for the detailed answer. I really appreciate this. 

> Unfortunately this is currently expected, since fully connected layers are currently not accelerated by larq compute engine. For more information please checkout #53

This makes sense, but I wonder why fully connected layers are left without binary implementation. Is there any specific reason for this? 

> There are a few subtleties here. `lq.models.summary()` gives you an estimate of the theoretical expected models size that is independent of the inference engine used.

So `lq.models.summary()` reports the size when all the parameters of the model are expressed in 1-bit right?

> * The summary might not be aware of all the fusion rules that the inference engine supports or might not recognise all model features correctly which means it could end up with a wrong estimate. We would be super happy for PRs improving these estimates in the model summary [here](https://github.com/larq/larq/blob/master/larq/models.py).

I can look into this. Are there any other issues related to this? 

> * In larq compute engine, binary weights are packed across the channel dimension into `int32` values so there might be some padding required that will increase the final model size compared to the theoretical estimate in cases where the number of filters are not multiples of 32. We have [docs for this available](https://docs.larq.dev/compute-engine/model_optimization_guide/) explaining these compute engine specifics in more detail.

From what I understood, the intuition behind this is that each `float32` weight is expressed as a single bit and since we have 32 filters we use the `int32` form to have one bit for each filter. Is this correct?

I am wondering whether what you have explained are available in docs somewhere, because the points that you listed are essential in implementing BNNs with Larq. From what I have understood, training BNNs is very easy with Larq but to have an optimal model in terms of memory and computational graph is more of an obscure art. There are so many nuances that one can get discouraged by the results and abandon BNNs entirely. Saying out loud what I have been observing in myself.

**Comment by AdamHillier** [comment]
Hi @rameshKrSah, thank you for your interest in Larq/LCE and you comments :)

> This makes sense, but I wonder why fully connected layers are left without binary implementation. Is there any specific reason for this?

This is on our feature wish-list, but so far there has not been strong demand for this. There aren't many literature models that use binary dense layers, for example. And as Lukas has already mentioned, there is the work around of using an (equivalent) 1x1 binary convolution instead, which does convert to a binary op in LCE.

That said, we would strongly welcome the efforts of any contributors who want to add binary dense layer support to LCE!

> So `lq.models.summary()` reports the size when all the parameters of the model are expressed in 1-bit right?

Not quite. The Larq model summary adds up the estimated size of the parameters for each layer. For binary layers (e.g. `QuantDense` or `QuantConv2D` with binary input/kernel quantisers) we calculate this size using 1-bit weights. For all others, we calculate this size using float32 weights. This is a bit of an inexact art though, because it can be hard to detect if a layer is binary or not (e.g. if custom quantisers are used), which is why the model summary should be seen as a rough guide only.

> From what I understood, the intuition behind this is that each `float32` weight is expressed as a single bit and since we have 32 filters we use the `int32` form to have one bit for each filter. Is this correct?

On the Larq/Keras side, everything (activations and weights) is floating point, and we simulate a binary network by rounding these values to binary — this is what the Larq binary quantisers do.

In the LCE converter, we take these weights, which are implemented in float but represent binary values, and perform what we refer to as 'bitpacking'. This means we compress each binary weight to a single bit. However, individual bits are not addressible in computer memory. We choose therefore to take 32 of these bits at a time and pack them together into a four-byte word. For this purpose, we use the type `int32` (four byte signed integer), but that's simply for convenience. This data is not a 32-bit integer, but 32 binary values inside one 32-bit integer wrapper.

The weights tensor is 4-dimensional, with shape `(out_channels, height, width, in_channels)`. We perform bitpacking along the innermost dimension, `in_channels`. Since we need to pack 32 bits at a time, this means that the input channels needs to be a multiple of 32. If that's not the case, we pad the input channel count with zeroes. This is why a weights tensor for a convolution with a single input channel takes up exactly the same amount of space as the weights tensor for a convolution with 32 input channels.

> From what I have understood, training BNNs is very easy with Larq but to have an optimal model in terms of memory and computational graph is more of an obscure art. There are so many nuances that one can get discouraged by the results and abandon BNNs entirely.

I am very sympathetic to this issue. Our goal with Larq and LCE is to provide tools to make BNN reasearch as easy as possible. The docs page [here](https://docs.larq.dev/compute-engine/model_optimization_guide/) that Lukas linked to is the place where we've tried to collate some of this information. However, our docs could always be better! If you have specific requests for specific things that you would like to be better explained in our docs, we'd be very happy to hear them.

**Comment by rameshKrSah** [comment]
Thank you for the the detailed explanation. It is much more clear to me now. I am closing this issue.
